### PR TITLE
Add a compiler warning for lowercase literals in patterns

### PR DESCRIPTION
**Pull request by dungpa**
UserVoice request (approved for F# 4.x): http://fslang.uservoice.com/forums/245727-f-language/suggestions/6668206-warn-when-literal-attribute-is-used-with-lowercase

This is a small and low-risk change to warn users when lowercase literals are ignored in favor of variable binding patterns.

Unit tests have been added.

### CODE REVIEW

**Comment by neoeinstein** [comment]
I like it, and it looks good from my review.

**Comment by dungpa** [comment]
@neoeinstein Thanks for reviewing. It's really useful.

**Comment by vasily-kirichenko** [comment]
Very important warning. Great job! :+1:

File: src/fsharp/TypeChecker.fs [hunk 1]
```diff
@@ -4700,7 +4700,14 @@ and TcPatBindingName _cenv _env id ty isMemberThis vis1 topValData (inlineFlag,d
     (fun (TcPatPhase2Input values) -> 
         let (vspec,typeScheme) = 
             match values.TryFind id.idText with
-            | Some x -> x
+            | Some value ->
+                let name = id.idText
+                if not (String.IsNullOrEmpty name) && Char.IsLower(name.[0]) then
```
Comment (zoren) [inline src/fsharp/TypeChecker.fs]: when would ```name``` be empty?

Comment (dungpa) [inline src/fsharp/TypeChecker.fs]: I'm not sure it will happen. The check is there to make sure that index access is totally safe.

Comment (zoren) [inline src/fsharp/TypeChecker.fs]: Ok me neither. However it does seem surprising that a identifier is the empty string.

**Comment by dsyme** [comment]
This looks good to me.

**Comment by KevinRansom** [comment]
committed.

**Comment by dungpa** [comment]
@KevinRansom I think it's better to reference the PR number and the issue number (if there is one) in the commit message. As it is committed now, it's really hard to blame the original author for bugs in the original PR :-).

**Comment by KevinRansom** [comment]
I screwed up the commit, I just reverted and am going to attribute it properly.
Sorry, it's kind a late here.

**Comment by PatrickMcDonald** [comment]
A quick look at the PR # explains everything!

**Comment by KevinRansom** [comment]
When I rebased, I forgot to set the author correctly. There are many times when I wished we just pressed the merge button.

Once again sorry for the brain-fart.  Now it's time for bed.

Kevin

**Comment by forki** [comment]
Dude, you are the one that is merging. Just make that decision to use the merge button. ;-)

**Comment by dungpa** [comment]
@KevinRansom No worries. I don't mind at all. The main problem is traceability and maintainability that we have to be careful with merging PRs.

**Comment by vasily-kirichenko** [comment]
@forki :+1:

**Pull request closed without merging**